cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -367,9 +372,180 @@ impl From<ProcessComposeLogContents> for LoggedError {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct ProcessComposeLogLine {
+    pub process: String,
+    pub message: String,
+}
+
+impl ProcessComposeLogLine {
+    /// Construct a new log line.
+    fn new(process: impl Into<String>, message: impl Into<String>) -> Self {
+        Self {
+            process: process.into(),
+            message: message.into(),
+        }
+    }
+}
+
+pub struct ProcessComposeLogStream {
+    readers: Vec<ProcessComposeLogReader>,
+    rx: Receiver<ProcessComposeLogLine>,
+}
+
+impl ProcessComposeLogStream {
+    /// Create a new log stream by attaching to the logs of multiple processes.
+    ///
+    /// For each `process` in `processes`, a new [ProcessComposeLogReader] will be started,
+    /// which will read log lines for the process and send them via MPSC to the receiver.
+    /// [ProcessComposeLogStream] implements [Iterator]
+    /// that will wait for log lines from any of the processes.
+    pub fn new(
+        socket: impl AsRef<Path>,
+        processes: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<ProcessComposeLogStream, ServiceError> {
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        let readers = processes
+            .into_iter()
+            .map(|process| ProcessComposeLogReader::start(socket.as_ref(), process, tx.clone()))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(ProcessComposeLogStream { readers, rx })
+    }
+}
+
+/// An iterator over log lines from multiple processes.
+///
+/// This iterator will block until a log line is received from any of the processes.
+/// Once _all_ processes have stopped,
+/// the iterator will return possible errors returned by the reader threads.
+/// Note that as long as at least one process is running,
+/// the iterator will keep outputting logs from that process.
+///
+/// Consider: send errors via the channel, to end logging early.
+impl Iterator for ProcessComposeLogStream {
+    type Item = Result<ProcessComposeLogLine, ServiceError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.rx.recv() {
+            Ok(line) => Some(Ok(line)),
+            // All senders have been dropped, so we wont't receive any more messages.
+            // Drain remaining reader return values.
+            Err(_) => {
+                loop {
+                    let reader: ProcessComposeLogReader = self.readers.pop()?;
```

Comment:
nit: can ignore if this is already obvious, might just be my own unfamiliarity with `?` returning None
```suggestion
                    // Returns None if there are no readers left
                    let reader: ProcessComposeLogReader = self.readers.pop()?;
```

Reply:
I think its clear in the context but, ill add the suggestion because here has a sepcific meaning to end the loop as well